pkg/prometheus/instance/instance.go

```diff
@@ -161,95 +158,82 @@ func New(globalCfg config.GlobalConfig, cfg Config, walDir string, logger log.Lo
-		for _, v := range i.cfg.ScrapeConfigs {
-			c[v.JobName] = v.ServiceDiscoveryConfig
-		}
-		i.exitErr = discoveryManagerScrape.ApplyConfig(c)
-		if i.exitErr != nil {
-			level.Error(i.logger).Log("msg", "failed applying config to discovery manager", "err", i.exitErr)
-			return
-		}
+// Run starts the instance and will run until an error happens during running
+// or until context was canceled.
+//
+// Run can be re-called after exiting.
+func (i *Instance) Run(ctx context.Context) error {
+	// Wrap our context so we can cancel ourselves.
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
+	defer cancel()
+
+	level.Debug(i.logger).Log("msg", "running instance", "name", i.cfg.Name)
+
+	// trackingReg wraps the register for the instance to make sure that if Run
+	// exits, any metrics Prometheus registers are removed and can be
+	// re-registered if Run is called again.
+	trackingReg := unregisterAllRegisterer{wrap: i.reg}
+	defer trackingReg.UnregisterAll()
+
+	wstore, err := i.newWal(&trackingReg)
+	if err != nil {
+		return err
 	}
 
-	remoteWriteConfigs := []*config.RemoteWriteConfig{}
-	for _, c := range i.cfg.RemoteWrite {
-		remoteWriteConfigs = append(remoteWriteConfigs, c.PrometheusConfig())
+	discoveryManager, err := i.newDiscoveryManager(ctx)
+	if err != nil {
+		return err
 	}
 
 	readyScrapeManager := &scrape.ReadyScrapeManager{}
-	remoteStore := remote.NewStorage(log.With(i.logger, "component", "remote"), &trackingReg, wstore.StartTime, i.walDir, i.cfg.RemoteFlushDeadline, readyScrapeManager)
-	i.exitErr = remoteStore.ApplyConfig(&config.Config{
-		GlobalConfig:       i.globalCfg,
-		RemoteWriteConfigs: remoteWriteConfigs,
-	})
-	if i.exitErr != nil {
-		level.Error(i.logger).Log("msg", "failed applying config to remote storage", "err", i.exitErr)
-		return
+	storage, err := i.newStorage(&trackingReg, wstore, readyScrapeManager)
+	if err != nil {
+		return err
 	}
 
-	fanoutStorage := storage.NewFanout(i.logger, wstore, remoteStore)
-
-	scrapeManager := scrape.NewManager(log.With(i.logger, "component", "scrape manager"), fanoutStorage)
-	i.exitErr = scrapeManager.ApplyConfig(&config.Config{
+	scrapeManager := scrape.NewManager(log.With(i.logger, "component", "scrape manager"), storage)
+	err = scrapeManager.ApplyConfig(&config.Config{
 		GlobalConfig:  i.globalCfg,
 		ScrapeConfigs: i.cfg.ScrapeConfigs,
 	})
-	if i.exitErr != nil {
-		level.Error(i.logger).Log("msg", "failed applying config to scrape manager", "err", i.exitErr)
-		return
+	if err != nil {
+		level.Error(i.logger).Log("msg", "failed applying config to scrape manager", "err", err)
+		return fmt.Errorf("failed applying config to scrape manager: %w", err)
 	}
-
 	readyScrapeManager.Set(scrapeManager)
 
-	level.Debug(i.logger).Log("msg", "creating host filterer", "for_host", i.hostname, "enabled", i.cfg.HostFilter)
-	filterer := NewHostFilter(i.hostname)
+	filterer, err := i.newHostFilter()
+	if err != nil {
+		return err
```

Comment:
I think you should wrap this error. Either here or in `newHostFilter()` to tell a better story if this somehow fails.